iree/tools/iree-benchmark-module-main.cc

```diff
@@ -165,9 +105,111 @@ void RegisterModuleBenchmarks(
       // we can make this setting configurable with a custom command line flag.
       ->Unit(benchmark::kMillisecond);
 }
+
+Status GetModuleContentsFromFlags(std::string& module_data) {
+  auto module_file = absl::GetFlag(FLAGS_module_file);
+  IREE_ASSIGN_OR_RETURN(module_data, file_io::GetFileContents(module_file));
+  return iree::OkStatus();
+}
+
+Status PrepareIREEVMResources(iree_vm_instance_t** instance,
+                              iree_hal_device_t** device,
+                              iree_vm_module_t** hal_module,
+                              iree_vm_context_t** context,
+                              iree_vm_module_t** input_module,
+                              const std::string& module_data) {
+  IREE_RETURN_IF_ERROR(iree_hal_module_register_types());
+  IREE_RETURN_IF_ERROR(
+      iree_vm_instance_create(iree_allocator_system(), instance));
+
+  // Create IREE's device and module.
+  IREE_RETURN_IF_ERROR(iree::CreateDevice(absl::GetFlag(FLAGS_driver), device));
+  IREE_RETURN_IF_ERROR(CreateHalModule(*device, hal_module));
+  IREE_RETURN_IF_ERROR(LoadBytecodeModule(module_data, input_module));
+
+  // Order matters. The input module will likely be dependent on the hal
+  // module.
+  std::array<iree_vm_module_t*, 2> modules = {*hal_module, *input_module};
+  IREE_RETURN_IF_ERROR(iree_vm_context_create_with_modules(
+      *instance, modules.data(), modules.size(), iree_allocator_system(),
+      context));
+  return iree::OkStatus();
+}
+
+Status RegisterSpecificFunction(iree_vm_instance_t* instance,
```

Comment:
probably worth throwing all this into a class so that you don't need to pass all the shared things around (instance/device/module/etc)